Sources/AblyChat/Logging.swift

```diff
@@ -1,14 +1,92 @@
     case error
     case silent
 }
+
+/// A reference to a line within a source code file.
+internal struct CodeLocation: Equatable {
+    /// A file identifier in the format used by Swift’s `#fileID` macro. For example, `"AblyChat/Room.swift"`.
+    internal var fileID: String
+    /// The line number in the source code file referred to by ``fileID``.
+    internal var line: Int
+}
+
+/// A log handler to be used by components of the Chat SDK.
+///
+/// This protocol exists to give internal SDK components access to a logging interface that allows them to provide rich and granular logging information, whilst giving us control over how much of this granularity we choose to expose to users of the SDK versus instead handling it for them by, say, interpolating it into a log message. It also allows us to evolve the logging interface used internally without introducing breaking changes for users of the SDK.
+internal protocol InternalLogger: Sendable {
+    /// Logs a message.
+    /// - Parameters:
+    ///   - message: The message to log.
+    ///   - level: The log level of the message.
+    ///   - codeLocation: The location in the code where the message was emitted.
+    func log(message: String, level: LogLevel, codeLocation: CodeLocation)
+}
+
+extension InternalLogger {
+    /// A convenience logging method that uses the call site’s #file and #line values.
+    public func log(message: String, level: LogLevel, fileID: String = #fileID, line: Int = #line) {
+        let codeLocation = CodeLocation(fileID: fileID, line: line)
+        log(message: message, level: level, codeLocation: codeLocation)
+    }
+}
+
+internal final class DefaultInternalLogger: InternalLogger {
+    // Exposed for testing.
+    internal let logHandler: LogHandler
+    internal let logLevel: LogLevel
+
+    internal init(logHandler: LogHandler?, logLevel: LogLevel?) {
+        self.logHandler = logHandler ?? DefaultLogHandler()
+        self.logLevel = logLevel ?? .info
+    }
+
+    internal func log(message: String, level: LogLevel, codeLocation: CodeLocation) {
+        guard level >= logLevel else {
+            return
+        }
+
+        // I don’t yet know what `context` is for (will figure out in https://github.com/ably-labs/ably-chat-swift/issues/8) so passing nil for now
+        logHandler.log(message: "(\(codeLocation.fileID):\(codeLocation.line)) \(message)", level: level, context: nil)
+    }
+}
+
+/// The logging backend used by ``DefaultInternalLogHandler`` if the user has not provided their own. Uses Swift’s `Logger` type for logging.
+internal final class DefaultLogHandler: LogHandler {
+    internal func log(message: String, level: LogLevel, context _: LogContext?) {
+        guard let osLogType = level.toOSLogType else {
+            // Treating .silent as meaning "don’t log it", will figure out the meaning of .silent in https://github.com/ably-labs/ably-chat-swift/issues/8
+            return
+        }
+
+        let logger = Logger()
```

Comment:
Is it ok to create it for each log line?

Reply:
`Logger` isn't marked as `Sendable` in the iOS 17.x SDKs, so it can't be stored as a property of a `Sendable` class. It does appear to be marked as `Sendable` in the iOS 18 SDK, so if we require the usage of Xcode 16 and above then I think we can change this. Created #40 to revisit.

Reply:
just a hunch but I imagine loggers to be similar to formatters (in that they're expensive) - swift concurrency is new to us all so not sure if this would even work (or if an actor would be ideal for us here) but there seems to be some way around it [here](https://forums.swift.org/t/non-sendable-type-logger-in-conformance-of-main-actor-isolated-property-logger-to-protocol-requirement-cannot-cross-actor-boundary/61802/2)

Reply:
I would suggest that we leave it as is for now and revisit in #40 once Xcode 16 is out — solutions that involve isolating the logger to an actor will have consequences such as having to make our logging method `async`, which then will affect the whole codebase including public API. (I think this might be a situation in which I’d prefer to just create an `@unchecked Sendable` type and manage access to the logger with a lock, but, as I say, I suggest we do nothing right now.)

Reply:
Sure thing!